tests/neg/OrType.scala

```diff
@@ -2,7 +2,7 @@ class B(val x: Int)
 class C(val x: Double)
 
 object Test{
-  def bar(x: B | C): Int | Double = x.x
+  def bar(x: B | C): Int | Double = x.x  // error
```

Comment:
I think it might help in some situations where people currently use structural types, for example https://github.com/lampepfl/dotty/issues/1175#issuecomment-242684995.
The proposed solution is to use a typeclass but if you assume that everyone should extend `java.io.Closeable` but still want to handle a few special cases that do not implement the interface but still follows its contract (like `scala.io.Source`) you could do:
```scala
type Closeable = java.io.Closeable | scala.io.Source
```